lib/src/room.dart

```diff
@@ -153,50 +159,18 @@ class Room {
   /// Adds the [state] to this room and overwrites a state with the same
   /// typeKey/stateKey key pair if there is one.
   void setState(Event state) {
-    // Decrypt if necessary
-    if (state.type == EventTypes.Encrypted && client.encryptionEnabled) {
-      try {
-        state = client.encryption?.decryptRoomEventSync(id, state) ?? state;
-      } catch (e, s) {
-        Logs().e('[LibOlm] Could not decrypt room state', e, s);
-      }
-    }
-
-    // We ignore room verification events for lastEvents
-    if (state.type == EventTypes.Message &&
-        state.messageType.startsWith('m.room.verification.')) {
-      return;
-    }
-
-    final isMessageEvent = {
-      EventTypes.Message,
-      EventTypes.Encrypted,
-      EventTypes.Sticker
-    }.contains(state.type);
-
-    // We ignore events editing events older than the current-latest here so
-    // i.e. newly sent edits for older events don't show up in room preview
-    final lastEvent = this.lastEvent;
-    if (isMessageEvent &&
-        state.relationshipEventId != null &&
-        state.relationshipType == RelationshipTypes.edit &&
-        lastEvent != null &&
-        !state.matchesEventOrTransactionId(lastEvent.eventId) &&
-        lastEvent.eventId != state.relationshipEventId &&
-        !(lastEvent.relationshipType == RelationshipTypes.edit &&
-            lastEvent.relationshipEventId == state.relationshipEventId)) {
-      return;
-    }
-
     // Ignore other non-state events
-    final stateKey = state.stateKey ??
-        (client.roomPreviewLastEvents.contains(state.type) ? '' : null);
+    final stateKey = state.stateKey;
     final roomId = state.roomId;
-    if (stateKey == null || roomId == null) {
+    if ((stateKey == null || roomId == null) &&
+        !client.importantStateEvents.contains(state.type)) {
+      Logs().w(
+        'Tried to set a non state event with type "${state.type}" as state event for a room',
+      );
       return;
     }
 
-    (states[state.type] ??= {})[stateKey] = state;
+    (states[state.type] ??= {})[stateKey ?? ''] = state;
```

Comment:
How would the stateKey still be null here? Isn't the check for the importantStateEvents wrong here? Since we want to store all state events but only state events?

Reply:
It can be null if the incoming state is of type `m.room.message`. We still need to support it here as a fallback. The legacy rooms still use this to determine the last event. The new lastEvent slot is only set once a new message comes in.

Yes, I could have written a database migration, but I had no motivation to do this for three databases at once. Once we have reduced them, we could refactor this but actually it does not hurt to have this here. The complexity is not that huge and we still use this logic to determine the last event if we have no other last event for a room

Reply:
But this is the setState call, couldn't we at least stop writing to it?

Reply:
no, this is called on client init when loading the room from the database. This way we load the lastEvent from the previous sdk versions to memory

Reply:
We should just set the last event then instead of writing it to the room state? And possibly only do that, when the last event is null? I guess that would have complications, if you have multiple events with a null state key, since this function doesn't know about that. But since we load events from the db into the room states, and that might include message events, we continue writing message events into the database and the room state on sync, which just seems confusing and not beneficial.

I guess the cleanest solution would be a database migration, but if you don't want to do that, at least document why this is still writing non-state events to the room state and what should be done about it in the future and maybe make the logic optional, if we already have a last event or so.

Reply:
We have multiple candidates for the lastEvent in the room state and the logic to pick the correct one is in the `Room.lastEvent` getter. We would need to completely rewrite this logic which would blow the PR even more up, if we want to set the lastEvent at this place, while we know that the legacy way is still working fine. Instead of a database migration we can just wait a little bit as the lastEvent picked by this logic is used to be stored in the database after the next sync for this room so this is somehow self healing already.

I can note this down there of course

Reply:
Added a comment